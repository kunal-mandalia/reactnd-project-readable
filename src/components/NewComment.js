import React, { Component } from 'react'
import PropTypes from 'prop-types'
import InlineEdit from './InlineEdit'
import { connect } from 'react-redux'
import { newCommentHide, createComment } from '../actions/index'
import { v1 as uuid } from 'uuid'
import '../styles/NewComment.css'
import '../styles/ActionBar.css'

export class NewComment extends Component {
  state = { body: '' }

  handleBodyChange = (e) => {
    this.setState({ body: e.target.value })
  }

  onSave = () => {
    if (this.state.body !== '') {
      const id = uuid()
      const timestamp = Date.now()
      const comment = {
        id,
        body: this.state.body,
        author: `theThingyBob`,
        parentId: this.props.parentId,
        timestamp
      }
      this.props.createComment(comment)
    }
  }

  render () {
    const { show, parentId } = this.props.newComment
    const showComment = show && parentId === this.props.parentId
    return (
      <div className={showComment ? 'new-comment' : 'hidden'}>
        <InlineEdit
          editMode={true}
          value={this.state.body}
          onChange={this.handleBodyChange}
          multiline
          autoFocus
          className={this.state.body !== '' ? 'edit-input-changed' : 'edit-input'}
        />
        <div className='action-bar'>
          [
            <a className={`action-bar-action ${this.state.body === '' ? 'disabled-link' : ''}`} onClick={this.onSave}> save </a> | 
            <a className='action-bar-action' onClick={() => { this.props.dispatch(newCommentHide()) }}> cancel </a>
          ]
        </div>
      </div>
    )
  }
}

const mapStateToProps = state => ({
  newComment: state.newComment
})

const mapDispatchToProps = dispatch => ({
  dispatch,
  createComment: comment => dispatch(createComment(comment))
})

export default connect(mapStateToProps, mapDispatchToProps)(NewComment)
