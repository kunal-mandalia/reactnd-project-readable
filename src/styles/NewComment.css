.new-comment {
  margin: 20px;
  padding-bottom: 20px;
}
