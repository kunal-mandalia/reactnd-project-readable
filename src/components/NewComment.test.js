import React from 'react'
import { NewComment } from './NewComment'
import { mount } from 'enzyme'
import renderer from 'react-test-renderer'

const mockFnOnSave = jest.fn()
const mockFnOnCancel = jest.fn()

const props = {
  parentId: `8tu4bsun805n8un48ve89`,
  onSave: mockFnOnSave,
  onCancel: mockFnOnCancel,
  newComment: {
    show: false,
    parentId: `8tu4bsun805n8un48ve89`
  }
}

describe(`<NewComment .../>`, () => {
  it(`should start hidden`, () => {
    const wrapper = mount(<NewComment {...props} />)
    expect(wrapper.find('.hidden')).toHaveLength(1)
  })

  it(`should match snapshot`, () => {
    const tree = renderer.create(<NewComment {...props} />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
