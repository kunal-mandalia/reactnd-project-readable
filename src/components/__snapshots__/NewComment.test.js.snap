// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewComment .../> should match snapshot 1`] = `
<div
  className="hidden"
>
  <textarea
    autoFocus={true}
    className="edit-input edit-input"
    cols="50"
    onChange={[Function]}
    placeholder=""
    rows="4"
    value=""
  />
  <div
    className="action-bar"
  >
    [
    <a
      className="action-bar-action disabled-link"
      onClick={[Function]}
    >
       save 
    </a>
     |
    <a
      className="action-bar-action"
      onClick={undefined}
    >
       cancel 
    </a>
    ]
  </div>
</div>
`;
